core/src/main/java/org/axonframework/eventsourcing/eventstore/FilteringDomainEventStream.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2010-2017. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.eventsourcing.eventstore;
+
+import java.util.function.Predicate;
+
+import org.axonframework.eventsourcing.DomainEventMessage;
+
+/**
+ * DomainEventStream implementation that filters a stream.
+ * 
+ * @author André Bierwolf
+ * @since 3.2.2
+ */
+public class FilteringDomainEventStream implements DomainEventStream {
+    private final DomainEventStream delegate;
+    private final Predicate<DomainEventMessage<?>> filter;
+    private Long lastSequenceNumber;
+
+    /**
+     * Initialize the stream, filter the given {@code stream} with the given
+     * {@code filter}.
+     *
+     * @param delegate The stream providing the elements
+     * @param filter   The filter to apply to the delegate stream
+     */
+    public FilteringDomainEventStream(DomainEventStream delegate, Predicate<DomainEventMessage<?>> filter) {
+        this.delegate = delegate;
+        this.filter = filter;
+    }
+
+    @Override
+    public DomainEventMessage<?> peek() {
+        if (!hasNext()) {
+            return null;
+        }
+        return delegate.peek();
```

Comment:
Do we want to apply the `filter` here?

Reply:
This is exactly the way it was implemented in Axon 2 
```java
 try {
                events = eventStore.readEvents(getTypeIdentifier(), aggregateIdentifier);
            } catch (EventStreamNotFoundException e) {
                throw new AggregateNotFoundException(aggregateIdentifier, "The aggregate was not found", e);
}
```
I assume there is a reason for this difference so that is why I added the filtering here with a possible performance penalty.

Reply:
I see, unfortunately, I'm not familiar with the reason for this. I guess that we should hear what @abuijze / @smcvb have to say on this.

Reply:
I assume we will eventually be returning the aggregate `type` as a criteria when event sourcing aggregates. Eventually being not a 3.x release, but a x.0 release.

The reason for not doing that right now is that adjusting this for all the users at this point might break backwards compatibility. Current users of Axon 3 might have for example adjusted the aggregate `type` between their private releases. As we're currently not querying on type, change the aggregate `type` (e.g. renaming the aggregate `class`) is fully allowed. It is hard to say whether this has occurred with any of our users of course.

We however fully embrace the idea to be backwards compatible between our versions.
As such, this functionality is fine, but I'd like to suggest a different approach to actually using it (as commented in the `EventSourcingRepository`)